sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -915,9 +916,9 @@ case class Cast(
     case TimestampType =>
       buildCast[Long](_, t => timestampToLong(t))
     case x: NumericType if ansiEnabled =>
-      b => x.exactNumeric.asInstanceOf[Numeric[Any]].toLong(b)
+      b => PhysicalNumericType.exactNumeric(x).toLong(b)
```

Comment:
One small issue with all of these changes is that we are doing the same resolution over and over on the hot path, while we only need to do it once. Can you move the resolution to the `PhysicalNumericType.numeric` out of the function, and do this for all calls in here?

For example:

``` scala
val exactNumeric = PhysicalNumericType.exactNumeric(x)
b => exactNumeric.toLong(b)
```

Reply:
Well after reading the code, I think we need to expland `NumericType` to all its sub-class and find the numeric directly to avoid `PhysicalNumericType.exactNumeric(x)`.

This is because we do not know the running type in advance.

Reply:
Essentially I think to achieve what you want to are doing something like this
```
case FloatType: PhysicalFloatType. exactNumeric()
case DoubleType: ....
...
```

This is already the least code implementation.

If we pull it out of the method it will even becomes this
```
Map map; 
case FloatType: map.put(FloatType, PhysicalFloatType. exactNumeric())

within methods
case FloatType: map,get(FloatType)
```
@hvanhovell what do you think?

Reply:
let's not overcomplicate it. The code here matches a data type and returns a lambda. We just need to get the `numeric` ahead of the lambda function body.

Reply:
Done